src/core/common/array_list.hpp

```diff
@@ -0,0 +1,153 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes the definition for array list class template.
+ */
+
+#ifndef ARRAY_LIST_HPP_
+#define ARRAY_LIST_HPP_
+
+#include <stddef.h>
+
+#include <openthread/error.h>
+
+#include "common/code_utils.hpp"
+
+namespace ot {
+template <typename Type> struct EqualTo
+{
+    bool operator()(const Type &lhs, const Type &rhs) const { return lhs == rhs; }
+};
+
+/**
+ * This class implements a list using array.
+ *
+ * @tparam Element   The type to element stored in the list.
+ * @tparam kMaxSize  The max size of the list.
+ *
+ */
+template <typename Element, ptrdiff_t kMaxSize> class ArrayList
```

Comment:
```suggestion
template <typename Element, uint16_t kMaxSize> class ArrayList
```
Can we use uint16_t ?

Reply:
Actually now I think it may not be a good idea to add this `ArrayList` to the codebase. It's more complicated than I thought. It should handle object destruction and copy control. Thoughts?

Reply:
Can we limit the object type to be POD types to simplify the implementation?

Reply:
I agree and also lean towards not adding `ArrayList` as its own data type (and use the model directly in posix RCP host code). 

I was considering if there would be other places within OT core where we could use this data type and could not immediately think of one. In general, I like the ["Three use to refactor" rule](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)).

Reply:
Removing this data structure seems good to me.

Reply:
Removed `ArrayList`.